geode-core/src/main/java/org/apache/geode/internal/cache/TXReservationMgr.java

```diff
@@ -123,23 +115,22 @@ private void release(IdentityArrayList localLocks, boolean conflictDetected) {
     for (int i = 0; i < llSize; i++) {
       TXRegionLockRequestImpl rr = (TXRegionLockRequestImpl) llArray[i];
       Object r = getRegionObject(rr);
-      Set keys = rr.getKeys();
-      Object curValue = this.regionLocks.get(r);
+      Map<Object, Boolean> keys = rr.getKeys();
+      Map<Object, Boolean> curValue = this.regionLocks.get(r);
```

Comment:
Also in releasing localLocks, you may need to track if "read" locks has been hold by this current transaction and whether need to decrement the read locks counter hold in the regionLocks by all transactions.
In existing code, no need to distinguish that -- as long as you foundIt, it is hold by this transaction. However, if conflictDetected, it could mean only part of the localLocks ("read" locks by this  transaction already put into the regionLocks) has been hold  needs to be released -- do not release "read" locks hold by other transactions.